### PR TITLE
LWDEV-9217 [Wamp] Support access token for authorization

### PR DESCRIPTION
**Pull request by MikhailGerasimov**

### CODE REVIEW

File: src/Lykke.Frontend.WampHost.Services/Security/ClientResolver.cs [hunk 1]
```diff
@@ -45,8 +50,14 @@ private string GetClientId(string token)
 
             try
             {
-                var sessionModel = _sessionService.GetAsync(token).Result;
-                clientId = sessionModel?.ClientId;
+                if (token.Length == LykkeTokenLength)
+                {
+                    clientId = _sessionService.GetAsync(token).GetAwaiter().GetResult()?.ClientId;
+                }
+                else
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/ClientResolver.cs]: I would replace "else" by "if(clientId == null)".

File: src/Lykke.Frontend.WampHost.Services/Security/PostConfigureOAuth2IntrospectionOptions.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            }
+            else
+            {
+                client = new DiscoveryClient(options.Authority);
+            }
+
+            client.Timeout = options.DiscoveryTimeout;
+            client.Policy = options.DiscoveryPolicy ?? new DiscoveryPolicy();
+            
+            var disco = await client.GetAsync().ConfigureAwait(false);
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/PostConfigureOAuth2IntrospectionOptions.cs]: disco?

File: src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            return string.IsNullOrWhiteSpace(value);
+        }
+
+        [DebuggerStepThrough]
+        public static bool IsPresent(this string value)
+        {
+            return !string.IsNullOrWhiteSpace(value);
+        }
+
+        internal static string Sha256(this string input)
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs]: This code is not used.

File: src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Lykke.Frontend.WampHost.Services.Security
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs]: I don't think this code has something to do with security.

File: src/Lykke.Frontend.WampHost/Lykke.Frontend.WampHost.csproj [hunk 5]
```diff
@@ -33,17 +33,20 @@
     </Content>
   </ItemGroup>
   <ItemGroup>
+    <None Update="appsettings.dev.json">
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost/Lykke.Frontend.WampHost.csproj]: We don't need it here.

File: src/Lykke.Frontend.WampHost/Modules/HostModule.cs [hunk 6]
```diff
@@ -10,6 +10,9 @@
 using Lykke.Frontend.WampHost.Services.Security;
 using Lykke.Service.ClientAccount.Client;
 using Lykke.Service.Session;
+using Microsoft.Extensions.Caching.Distributed;
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost/Modules/HostModule.cs]: I suppose lines 13-15 are not used.

File: src/Lykke.Frontend.WampHost.Services/Security/AsyncLazy.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Lykke.Frontend.WampHost.Services.Security
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/AsyncLazy.cs]: I don't think this code has something to do with security.

File: src/Lykke.Frontend.WampHost.Services/Security/ClaimConverter.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Security.Claims;
+using Newtonsoft.Json;
+
+namespace Lykke.Frontend.WampHost.Services.Security
+{
+    internal class ClaimConverter : JsonConverter
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/ClaimConverter.cs]: This class is not used.

File: src/Lykke.Frontend.WampHost.Services/Security/OAuth2IntrospectionOptions.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        /// <summary>
+        /// Specifies the HTTP handler for the introspection endpoint
+        /// </summary>
+        public HttpMessageHandler IntrospectionHttpHandler { get; set; }
+
+
+        /// <summary>
+        /// Specifies the method how to retrieve the token from the HTTP request
+        /// </summary>
+        public Func<HttpRequest, string> TokenRetriever { get; set; } = TokenRetrieval.FromAuthorizationHeader();
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/OAuth2IntrospectionOptions.cs]: I don't see usages of this code.

File: src/Lykke.Frontend.WampHost.Services/Security/TokenRetrieval.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    /// Defines some common token retrieval strategies
+    /// </summary>
+    internal static class TokenRetrieval
+    {
+        /// <summary>
+        /// Reads the token from the authrorization header.
+        /// </summary>
+        /// <param name="scheme">The scheme (defaults to Bearer).</param>
+        /// <returns></returns>
+        public static Func<HttpRequest, string> FromAuthorizationHeader(string scheme = "Bearer")
```
Comment (polonzer) [inline src/Lykke.Frontend.WampHost.Services/Security/TokenRetrieval.cs]: I don't see usages of this code.

File: src/Lykke.Frontend.WampHost.Services/Security/ClientResolver.cs (hunk 1 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/ClientResolver.cs]: These two cases are mutual excluded. If one of them false we don't need to check the other.

File: src/Lykke.Frontend.WampHost.Services/Security/PostConfigureOAuth2IntrospectionOptions.cs (hunk 2 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/PostConfigureOAuth2IntrospectionOptions.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs (hunk 3 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs (hunk 4 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/StringExtensions.cs]: Fixed

File: src/Lykke.Frontend.WampHost/Lykke.Frontend.WampHost.csproj (hunk 5 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost/Lykke.Frontend.WampHost.csproj]: Fixed

File: src/Lykke.Frontend.WampHost/Modules/HostModule.cs (hunk 6 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost/Modules/HostModule.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/AsyncLazy.cs (hunk 7 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/AsyncLazy.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/ClaimConverter.cs (hunk 8 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/ClaimConverter.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/OAuth2IntrospectionOptions.cs (hunk 9 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/OAuth2IntrospectionOptions.cs]: Fixed

File: src/Lykke.Frontend.WampHost.Services/Security/TokenRetrieval.cs (hunk 10 above)
Comment (MikhailGerasimov) [reply src/Lykke.Frontend.WampHost.Services/Security/TokenRetrieval.cs]: Fixed

**Pull request merged**